docs/developer/ceed.md

```diff
@@ -0,0 +1,206 @@
+- The use libCEED allows RDycore to compute the RHS-function for the explicit time-integrators on CPU or GPU, which can also be selected at runtime via
+     - For CPUs: `-ceed /cpu/self`
+     - For NVIDIA GPUs: `-ceed /gpu/cuda -dm_vec_type cuda`
+     - For AMD GPUs: `-ceed /gpu/hip -dm_vec_type hip`
+
+## Example Mesh
+
+Let's consider an example 3x2 mesh that consists of
+
+- 6 cells: `c00` to `c05`
+- 17 edges: `e00` to `e16`
+    - Internal edges: `e04`, `e05`, `e07`, `e08`, `e09`, `e11`, and `e12`.
+    - Boundary edges: `e00`, `e01`, `e02`, `e03`, `e06`, `e10`, `e13`, `e14`, `e15`, and `e16`.
+- 12 vertices: `v00` to `v11`
+
+```text
+v08---e14---v09---e15---v10---e16---v11
+ |           |           |           |
+ |           |           |           |
+e10   c03   e11   c04   e12   c05   e13
+ |           |           |           |
+ |           |           |           |
+v04---e07---v05---e08---v06---e09---v07
+ |           |           |           |
+ |           |           |           |
+e03   c00   e04   c01   e05   c02   e06
+ |           |           |           |
+ |           |           |           |
+v00---e00---v01---e01---v02---e02---v03
+
+```
+
+For the mesh shown above, the prognostic variables of the 2D SWE are saved in a strided PETSc Vec (`X`).
+The block size of `X` is `3` corresponding to the following prognostic variables:
+
+- Height (`h`),
+- Momentum in x-dir (`hu`), and
+- Momentum in y-dir (`hv`).
+The size of `X` will `6 * 3` where `6` corresponds to number of cells in the mesh. The layout of `X` will be as follows:
+
+```text
+X = [[h0 hu0 hv0] [h1 hu1 hv1] ... [h5 hu5 hv5]]
+```
+
+## SWE physics with libCEED
+
+The libCEED version of RDycore's explicit time-integrator of the SWE solver has two `CeedOperator`:
+
+1. `rdy->ceed_rhs.op_edges` : Computes fluxes across edges and it includes two sub-operators that correspond to:
+    - Internal edges,
+    - Boundary edges
+
+2. `rdy->ceed_rhs.src`: Opeator that computes the source terms including rainfall and terms associated with bed slope and bed friction.
+
+### `CeedOperator` for Internal Edges
+
+RDycore uses first-order finite volume discretization to compute the flux across the edges, which requires values on the left and the right of the edge.
+
+| Internal Edge | Left Cell | Right cell |
+| ------------- | --------- | ---------- |
+| e04           | c00       | c01        |
+| e05           | c01       | c02        |
+| e07           | c00       | c03        |
+| e08           | c01       | c04        |
+| e09           | c02       | c05        |
+| e11           | c03       | c04        |
+| e12           | c04       | c05        |
+
+The steps involved in creating the `CeedOperator` associated with the internal edges are as follows:
+
+- First, create a `CeedQFunction`, `qf`, and add input and output fields.
+The input fields includes geometric attributes associated with the edges and the prognostic variables left and right of the edges.
```

Comment:
includes -> include
left and right of the edges -> to the left and right of the edges